tests/test_lion8b.py

```diff
@@ -0,0 +1,422 @@
+_FLOAT_DTYPES = [torch.bfloat16, torch.float16, torch.float32]
+
+np.set_printoptions(linewidth=160, formatter={'float': lambda f: f'{f:5.3f}'})
+
+
+@pytest.mark.gpu
+@pytest.mark.parametrize('N,D', _MANY_PARAM_SHAPES)
+@pytest.mark.parametrize('dtype', _FLOAT_DTYPES)
+@pytest.mark.parametrize('fused,use_errors', [(False, False), (True, False),
+                                              (True, True)])
+def test_modifies_weights_and_momentums(N: int, D: int, dtype: torch.dtype,
+                                        fused: bool, use_errors: bool) -> None:
+    device = 'cuda'
+    torch.manual_seed(123)
+    X = torch.randn((N, D), device=device, requires_grad=False, dtype=dtype)
+    W = torch.randn((D, D), device=device, requires_grad=True, dtype=dtype)
+    W_orig = W.detach().clone()
+
+    opt = Lion8bit([W],
+                   lr=1.0,
+                   _fused=fused,
+                   betas=(.75, .75),
+                   weight_decay=.2,
+                   error_correction=use_errors)
+
+    Y = X @ W
+    loss = Y.sum()
+    loss.backward()
+    torch.testing.assert_close(W_orig, W)  # no weight modification yet
+    opt.step()
+    opt.zero_grad()
+
+    with pytest.raises(AssertionError):  # opt step modified the weights
+        torch.testing.assert_close(W_orig, W)
+
+    # every momentum should be nonzero with infinite precision, but
+    # might be zero after quantization
+    param_state = opt.state[W]  # type:ignore using tensor as key
+    momentum = param_state['exp_avg'].materialize()
+    assert momentum.shape == (D, D)
+    momentum = momentum.ravel()
+    assert momentum is not None
+    if momentum.numel() == 1:
+        assert momentum.item() != 0
+    else:
+        assert torch.std(momentum).item() > 0
+
+
+@pytest.mark.gpu
+@pytest.mark.parametrize('N,D', _MANY_PARAM_SHAPES)
+@pytest.mark.parametrize('device', ['cpu', 'cuda'])
+@pytest.mark.parametrize('dtype', _FLOAT_DTYPES)
+@pytest.mark.parametrize('weight_decay', [0, .1])
+@pytest.mark.parametrize('fused,use_errors', [(False, False), (True, False),
+                                              (True, True)])
+def test_changes_with_zero_grads(N: int, D: int, device: str,
+                                 dtype: torch.dtype, weight_decay: float,
+                                 fused: bool, use_errors: bool) -> None:
+    if (dtype != torch.float32) and device == 'cpu':
+        return
+    torch.manual_seed(123)
+    W = torch.rand((D, D), device=device, requires_grad=True)
+    with torch.no_grad():
+        W += torch.sign(W)  # bound away from zero so decay won't change sign
+    W_orig = W.detach().clone()
+
+    opt = Lion8bit([W],
+                   _fused=fused,
+                   betas=(.5, .5),
+                   quantize=(device != 'cpu'),
+                   weight_decay=weight_decay,
+                   error_correction=use_errors)
+
+    zeros_grad = torch.zeros_like(W)
+    for _ in range(5):
+        W.grad = zeros_grad
```

Comment:
```suggestion
    for _ in range(5):
        W.grad.zeros_()
```

https://pytorch.org/docs/stable/generated/torch.Tensor.zero_.html#torch.Tensor.zero_

Reply:
So I started to change this, but then I then I realized it's like this because torch optimizers' `zero_grad` [defaults to setting the grad to None now](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html); this means we actually do have to assign to `.grad` since `W.grad == None`. Or we could `zero_grad(set_to_none=False)`, but I'd rather test the default behavior.